pkg/manager/member/pvc_cleaner.go

```diff
@@ -40,37 +47,144 @@ type PVCCleanerInterface interface {
+
+	return rpc.reclaimPV(tc)
+}
+
+func (rpc *realPVCCleaner) reclaimPV(tc *v1alpha1.TidbCluster) (map[string]string, error) {
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
+
 	// for unit test
 	skipReason := map[string]string{}
 
-	selector, err := label.New().Instance(tc.GetLabels()[label.InstanceLabelKey]).Selector()
+	pvcs, err := rpc.listAllPVCs(tc)
 	if err != nil {
-		return skipReason, fmt.Errorf("cluster %s/%s assemble label selector failed, err: %v", ns, tcName, err)
+		return skipReason, err
 	}
 
-	pvcs, err := rpc.pvcLister.PersistentVolumeClaims(ns).List(selector)
+	for _, pvc := range pvcs {
+		pvcName := pvc.GetName()
+		l := label.Label(pvc.Labels)
+		if !(l.IsPD() || l.IsTiKV()) {
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotPDOrTiKV
+			continue
+		}
+
+		if pvc.Status.Phase != corev1.ClaimBound {
+			// If pvc is not bound yet, it will not be processed
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotBound
+			continue
+		}
+
+		if len(pvc.Annotations[label.AnnPVCDeferDeleting]) == 0 {
+			// This pvc has not been marked as defer delete PVC, can't reclaim the PV bound to this PVC
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotDeferDeletePVC
+			continue
+		}
+
+		// PVC has been marked as defer delete PVC, try to reclaim the PV bound to this PVC
+		podName, exist := pvc.Annotations[label.AnnPodNameKey]
+		if !exist {
+			// PVC has not pod name annotation, this is an unexpected PVC, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotHasPodNameAnn
+			continue
+		}
+
+		_, err := rpc.podLister.Pods(ns).Get(podName)
+		if err == nil {
+			// PVC is still referenced by this pod, can't reclaim PV
+			skipReason[pvcName] = skipReasonPVCCleanerPVCeferencedByPod
+			continue
+		}
+		if !errors.IsNotFound(err) {
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pod %s from local cache failed, err: %v", ns, tcName, pvcName, podName, err)
+		}
+
+		// if pod not found in cache, re-check from apiserver directly to make sure the pod really not exist
+		_, err = rpc.kubeCli.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
+		if err == nil {
+			// PVC is still referenced by this pod, can't reclaim PV
+			skipReason[pvcName] = skipReasonPVCCleanerPVCeferencedByPod
+			continue
+		}
+		if !errors.IsNotFound(err) {
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pod %s from apiserver failed, err: %v", ns, tcName, pvcName, podName, err)
+		}
+
+		// Without pd or tikv pod reference this defer delete PVC, start to reclaim PV
+		pvName := pvc.Spec.VolumeName
+		pv, err := rpc.pvLister.Get(pvName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				skipReason[pvcName] = skipReasonPVCCleanerNotFoundPV
+				continue
+			}
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pv %s failed, err: %v", ns, tcName, pvcName, pvName, err)
+		}
+
+		if pv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimDelete {
+			skipReason[pvcName] = skipReasonPVCCleanerPVWithDeletePolicy
```

Comment:
Should we skip here? If the reclaim policy of the PV was patched to `delete` and the deletion to PVC was failed in the last round of reconcile, the cleaner would never be able to reclaim the PV.

Reply:
Thank you for pointing this out, this is a bug.